ircd/send.c

```diff
@@ -1503,3 +1503,163 @@ kill_client_serv_butone(struct Client *one, struct Client *target_p, const char
+	multiline_item_start = multiline_prefix + multiline_prefix_len;
+	multiline_cur = multiline_item_start;
+	multiline_cur_len = multiline_prefix_len;
+	multiline_remote_extra_space = 0;
+	return true;
+}
+
+bool
+sendto_one_multiline_remote_extra_space(struct Client *target_p, struct Client *client_p)
+{
+	ssize_t remote_extra_space;
+
+	if (target_p != multiline_stashed_target_p)
+	{
+		s_assert(false && "Multiline: missed init call!");
+		multiline_stashed_target_p = NULL;
+		return false;
+	}
+
+	remote_extra_space = strlen(client_p->name) - strlen(client_p->id);
+
+	if (remote_extra_space > 0)
+	{
+		multiline_remote_extra_space += remote_extra_space;
+	}
+
+	return true;
+}
+
+enum multiline_item_result
+sendto_one_multiline_item(struct Client *target_p, const char *format, ...)
+{
+	va_list args;
+	char item[DATALEN];
+	int item_len, res;
+	enum multiline_item_result ret = MULTILINE_SUCCESS;
+
+	if (target_p != multiline_stashed_target_p)
+	{
+		s_assert(false && "Multiline: missed init call!");
+		multiline_stashed_target_p = NULL;
+		return MULTILINE_FAILURE;
+	}
+
+	va_start(args, format);
+	item_len = vsnprintf(item, sizeof item, format, args);
+	va_end(args);
+
+	if (item_len < 0 || item_len >= DATALEN || multiline_prefix_len + multiline_remote_extra_space + item_len > DATALEN)
+	{
+		s_assert(false && "Multiline: failure preparing item!");
+		multiline_stashed_target_p = NULL;
+		return MULTILINE_FAILURE;
+	}
+
+	if (multiline_cur_len + item_len > DATALEN - multiline_remote_extra_space)
+	{
+		/* strip trailing separator */
+		multiline_cur[-multiline_separator_len] = '\0';
+		sendto_one(target_p, "%s", multiline_prefix);
+		multiline_cur_len = multiline_prefix_len;
+		multiline_cur = multiline_item_start;
+		ret = MULTILINE_WRAPPED;
+	}
+
+	res = snprintf(multiline_cur, sizeof multiline_prefix - (multiline_cur_len - multiline_prefix_len), "%s%s", item, multiline_separator);
+
+	if (res < 0)
+	{
+		s_assert(false && "Multiline: failure appending item!");
+		multiline_stashed_target_p = NULL;
+		return MULTILINE_FAILURE;
+	}
+
+	multiline_cur_len += res;
+	multiline_cur += res;
+	return ret;
+}
+
+bool
+sendto_one_multiline_fini(struct Client *target_p, const char *format, ...)
+{
+	va_list args;
+	char final[DATALEN];
+	int final_len;
+
+	if (target_p != multiline_stashed_target_p)
+	{
+		s_assert(false && "Multiline: missed init call!");
+		multiline_stashed_target_p = NULL;
+		return false;
+	}
+
+	if (multiline_cur_len == multiline_prefix_len)
+	{
+		multiline_stashed_target_p = NULL;
+		return true;
+	}
+
+	if (format)
+	{
+		va_start(args, format);
+		final_len = vsnprintf(final, sizeof final, format, args);
+		va_end(args);
+
+		if (final_len < 0 || final_len >= DATALEN || final_len > multiline_prefix_len)
+		{
+			s_assert(false && "Multiline: failure preparing final prefix!");
+			multiline_stashed_target_p = NULL;
+			return false;
+		}
+	}
+	else
+	{
+		rb_strlcpy(final, multiline_prefix, sizeof final);
```

Comment:
This adds an extra copy of the buffer before `multiline_item_start`.

```suggestion
		rb_strlcpy(final, multiline_prefix, multiline_prefix_len + 1);
```

Reply:
This was fine until *you* wanted to eliminate the separate items buffer 😛